### DISCUSSION TITLE
Adavanced title blocks editing

### DISCUSSION THREAD

**Discussion by Cicorione** [speedata/publisher#496] [Q&A] [answered]
Dear All,

I have been studying the documentation but I couldn't find any direct information for my problem. Personally this is my most blocking issue, but I need to accelerate since my bosses are waiting for my demonstration and I need to have it ready ASAP...

My bosses like designing the title blocks this way:

![sd-titles](https://github.com/speedata/publisher/assets/19494201/a9a5c747-5a16-48bd-be39-72341e569bf8)

So far I have been only came out, theoretically, with the idea to handle these titles with a simple table, as in the ancient HTML4 times; thus the side cells have the rounded border and any graphics which is not a vertical straight line is another cell, and between them you have the text you need to display. This is not elegant and perhaps pretty convoluted but it is the only solution I could elaborate.

Also in GUI Interact Editors you can work with tabulations and alignments to make text stick to a specific distance or position:

![sd-tab](https://github.com/speedata/publisher/assets/19494201/45bd86fe-bb51-40fe-9dab-311f5870c90d)

In SP you may play with the cursors as explained here for: https://doc.speedata.de/publisher/en/basics/positioningarea/#ch-cursor ; however the chance that I got it totally wrong are very high!

By the way I was able to get a similar title block playing with ConTeXt, which has specific tools to handle text box areas, however Metafun is slightly easier to understand than Metapost. The manual says that SP has limited support for Metfun though. Long-story short I couldn't find any specific instruction or command to achieve a result that is enough close to the title blocks above, Honestly I would avoid to recur to any dirty tricks as the old way to do the buttons always in HTML4.

Perhaps there is some depth documentation separated from the regular manual that I can attempt to study, at least trying... 
Perhaps with pure Metpost, which is not fun, I can try (if I am able) to achieve something. I don't have any ideas... 

Please any suggestion such as further documentation to read, ideas, codes, will be really appreciated, thanks! 🙏

**Comment by pgundlach** [comment, marked as answer]
What you want is (as far as I understand) currently not possible with pure speedata Publisher: it does not support color shading / gradients. Nor does it support drop shadows on letters.

That said, if supposedly sp does gradients with MetaFun, I'd use a table:

```xml
<Layout xmlns="urn:speedata.de:2009/publisher/en"
    xmlns:sd="urn:speedata:2009/publisher/functions/en">

    <DefineGraphic name="blue">
        filldraw box withcolor blue ;
    </DefineGraphic>
    <DefineGraphic name="new">
        filldraw  (0,0) -- (box.width,0) -- (box.width, box.height) --
        (box.width, box.height) -- (-0.5cm, box.height ) -- cycle  withcolor 0.7red ;
    </DefineGraphic>


    <DefineGraphic name="bluerounded">
        filldraw box withcolor blue ;
        clip currentpicture to roundedsquare(box.width + 7mm, box.height, 3mm);
    </DefineGraphic>
    <DefineGraphic name="newrounded">
        filldraw  (0,0) -- (box.width,0) -- (box.width, box.height) --
        (box.width, box.height) -- (-0.5cm, box.height ) -- cycle  withcolor 0.7red ;
        clip currentpicture to roundedsquare(box.width + 1cm, box.height, 3mm) shifted (-1cm,0mm);
    </DefineGraphic>

    <!-- first start with a table -->
    <Record element="data">
        <PlaceObject>
            <Table padding="6pt" width="20cm" >
                <Tr>
                    <Td padding-right="1cm"><Paragraph><Value>EXTENSION CABLE LED WIRING</Value></Paragraph></Td>
                    <Td><Paragraph><Value>NEW</Value></Paragraph></Td>
                </Tr>
            </Table>
        </PlaceObject>

        <NextRow />

        <!-- add text color and graphic -->
        <PlaceObject>
            <Table padding="6pt" width="20cm" >
                <Tr>
                    <Td graphic="blue" padding-right="1cm"><Paragraph color="white"><Value>EXTENSION CABLE LED WIRING</Value></Paragraph></Td>
                    <Td graphic="new"><Paragraph color="white"><Value>NEW</Value></Paragraph></Td>
                </Tr>
            </Table>
        </PlaceObject>

        <NextRow />

        <!-- with rounded corners - same table, just different graphics-->
        <PlaceObject>
            <Table padding="6pt" width="20cm" >
                <Tr>
                    <Td graphic="bluerounded" padding-right="1cm"><Paragraph color="white"><Value>EXTENSION CABLE LED WIRING</Value></Paragraph></Td>
                    <Td graphic="newrounded"><Paragraph color="white"><Value>NEW</Value></Paragraph></Td>
                </Tr>
            </Table>
        </PlaceObject>
    </Record>
</Layout>
```
<img width="670" alt="table" src="https://github.com/speedata/publisher/assets/209434/084f7ba3-ac1f-4076-ab0c-428eefad5526">

So there are three tables:

The first is a simple table with a bit of padding to get sizing correct (there are other means, like setting the column width).

The next table switches the text color to white then adds the metapost graphics. The left cell is straightforward, the “new” cell has the top left corner shifted.

The third table clips the current picture to a rounded rectangle.

So the problem of the gradient and the font shadow remains. Both can be implemented in the speedata Publisher. (These are not trivial patches though).

**Reply by Cicorione** [reply]
Thank you very much...

Eventually I was on the right track as well.
It is the same old trick to put a background image into a cell table, for this scope perhaps I don't need to use Metapost exclusively, I may use also a graphic made with another software. I remember that ConTeXt has specific macros to handle text boxes such as `/framed`, and I was able (everything pretty easy for me is a great challenge) to create a macro that overlapped two frame round boxes to create a similar title block. Perhaps also Speedata has such feature.

I don't say anything special truth here but gradients and drop shadows are common tools for any designer, and not having them is a real limitation, I hope this will be integrated as soon as possible; said that I would have other two questions which are related but different:

1. I am trying to to put my items into a grid, I'd like to use the grid to define how many products go into a spread, let's say a specific spread contain 12 products, then I would have a page grid of 2x3. Now I want have control on these title blocks ad say that those must be 80% of the grid cell width. Reading the documentation I noticed there are some macros that begin with `sp:some-command` that have control over these parameters, however I couldn't find a complete list of these macros with their full explanation. Please could you redirect me to such documentation?

2. The second question is: only few products are newer, and only them need to be highlighted as new — actually other products needs further categorization but I guess the procedure would, be more or less, the same — thus I need to define on `data.xnl` who is new and on the `layout.xml` to define some conditions that say when an item from `data.xml` is marked as `new` then use this rule. Is that possible?

Thanks... 🙏

**Reply by pgundlach** [reply]
>  I remember that ConTeXt has specific macros to handle text boxes such as `/framed`, and I was able (everything pretty easy for me is a great challenge) to create a macro that overlapped two frame round boxes to create a similar title block. Perhaps also Speedata has such feature.

not sure what you exactly need, but you can put a [background image](https://doc.speedata.de/publisher/en/tables2/#_background_in_table_rows) in a table cell.


> I don't say any special truth but gradients and drop shadows are common tools for any designer, and not having them is a real limitation, I hope this will be integrated as soon as possible,

I have created catalogs since more than 14 years now and never had the (paid) request for gradients, so it might be common, but never occurred to me. Sometimes I use pre-generated PDF files for background images for that purpose.

> 
> 1. I am trying to to put my items into a grid, I'd like to use the grid to define how many products go into a spread, let's say a specific spread contain 12 products, then I would have a page grid of 2x3. 

Yes, that would be the grid of choice.


> Now I want have control on these title blocks ad say that those must be 80% of the grid cell width. Reading the documentation I noticed there are some macros that begin with `sp:some-command` that have control over these parameters, however I couldn't find a complete list of these macros with their full explanation. Please could you redirect me to such documentation?

Did you see the [list of layout functions](https://doc.speedata.de/publisher/en/xpathfunctions/#_the_following_xpath_functions_are_known_to_the_system)?

depending on what you need a table could do what you want:

```xml
<Layout xmlns="urn:speedata.de:2009/publisher/en"
    xmlns:sd="urn:speedata:2009/publisher/functions/en">
    <SetGrid nx="2" ny="3"/>
    <Record element="data">
        <ForAll select="item">
            <PlaceObject>
                <Table width="1" stretch="max">
                    <Columns>
                        <Column width="10pt" />
                        <Column width="1*"/>
                    </Columns>
                    <Tablerule />
                    <Tr>
                        <Td colspan="2">
                            <Switch>
                                <Case test="@new='true'">
                                    <Paragraph><Value>NEW!!</Value></Paragraph>
                                </Case>
                            </Switch>
                        </Td>
                    </Tr>
                    <Tr>
                        <Td>
                            <Paragraph>
                                <Value select="position()"></Value>
                            </Paragraph>
                        </Td>

                        <Td>
                            <Paragraph>
                                <Value select="@text"></Value>
                            </Paragraph>
                        </Td>
                    </Tr>

                </Table>
            </PlaceObject>
        </ForAll>
    </Record>
</Layout>
```

with the following data file:

```xml
<data>
  <item new="false" text="an item" />
  <item new="false" text="an item" />
  <item new="true" text="an item" />
  <item new="true" text="an item" />
  <item new="false" text="an item" />
  <item new="false" text="an item" />
  <item new="false" text="an item" />
</data>
```

But you can also define a Group with a smaller grid and position the Group into one of these grid

You can also use a two step workflow: first step is to create small pages (1/12 of the page size) and include these as PDF pages.

So it really depends on the layout you want to achieve. If you can make a mockup, I'd be happy to share my thoughts.



> 2. The second question is: only few products are newer, and only them need to be highlighted as new — actually other products needs further categorization but I guess the procedure would, be more or less, the same — thus I need to define on `data.xnl` who is new and on the `layout.xml` to define some conditions that say when an item from `data.xml` is marked as `new` then use this rule. Is that possible?

See the XML above. Does this answer the question?

**Reply by Cicorione** [reply]
Thank you very much.

I'll try to put all these information together, create a demo, and ask for a paid subscription. Our main catalog has around 3500 items, and I'd like to move everything away from any Adobe software.  However I believe that without a professional support this task would be impossible to achieve. I need to do a perfect presentation, but I am juggling with my day to day tasks and deadline, perhaps when ready I can have you reviewing my code.

Thanks again!

**Reply by pgundlach** [reply]
Don't hesitate to ask. The subscription is the last part before going live with a catalog. There is always the possibility to download the Pro software and getting help from me. Seriously, don't be afraid, I am here to help (you can find the [mail address on the web page](https://www.speedata.de/imprint/))

I will always honestly say what is easily possible and what is difficult to achieve.
